src/Distribution/Nixpkgs/Haskell/Platform.hs

```diff
@@ -0,0 +1,123 @@
+  "watchos"       -> "ios"
+  "tvos"          -> "ios"
+  "linux-android" -> "linux-android"
+  "linux-androideabi" -> "linux-androideabi"
+  _ | "linux-" `isPrefixOf` os -> "linux"
+  _ -> fromMaybe os $ listToMaybe
+    [prefix | prefix <- osPrefixes, prefix `isPrefixOf` os]
+  where osPrefixes =
+          [ "gnu", "openbsd", "aix", "darwin", "solaris2", "freebsd", "nto-qnx"]
+
+parseArch :: String -> Arch
+parseArch = classifyArch Permissive . ghcConvertArch
+
+parseOS :: String -> OS
+parseOS = classifyOS Permissive . ghcConvertOS
+
+parsePlatformParts :: [String] -> Maybe Platform
+parsePlatformParts = \case
+  [arch, os] ->
+    Just $ Platform (parseArch arch) (parseOS os)
+  (arch : _ : osParts) ->
+    Just $ Platform (parseArch arch) $ parseOS $ intercalate "-" osParts
+  _ -> Nothing
+
+-- | Convert a platform string of two or three(-ish) components to 'Platform'.
+--
+--   For this, the following logic is utilized:
+--
+--   - If the string has one dash, the form @cpu-os@ is assumed where @os@ may
+--     only have a single component. The @vendor@ part is ignored.
+--
+--   - Otherwise @cpu-vendor-os@ is assumed where @os@ may have any number of
+--     components separated by dashes to accomodate its two component
+--     @kernel-system@ form.
+--
+--   __Note:__ This behavior is different from nixpkgs' @lib.systems.elaborate@:
+--   Because we have no knowledge of the legal contents of the different parts,
+--   we only decide how to parse it based on what form the string has. This can
+--   give different results compared to autoconf or nixpkgs. It will also never
+--   reject an invalid platform string that has a valid form.
+--
+--   >>> parsePlatformLenient "x86_64-unknown-linux"
+--   Just (Platform X86_64 Linux)
+--   >>> parsePlatformLenient "x86_64-pc-linux-gnu"
+--   Just (Platform X86_64 Linux)
+--   >>> parsePlatformLenient "x86_64-linux"
+--   Just (Platform X86_64 Linux)
+--
+--   __Note__ also that this conversion sometimes looses information nixpkgs
+--   would retain:
+--
+--   >>> parsePlatformLenient "powerpc64-unknown-linux"
+--   Just (Platform PPC64 Linux)
+--   >>> parsePlatformLenient "powerpc64le-unknown-linux"
+--   Just (Platform PPC64 Linux)
+parsePlatformLenient :: String -> Maybe Platform
+parsePlatformLenient = parsePlatformParts . splitOn "-"
+
+-- | Convert a Nix style system tuple into a Cabal 'Platform'. The tuple is
+--   assumed to be of the form @cpu-os@, any extra components are assumed to be
+--   part of @os@ to accomodate its @kernel-system@ form.
+--
+--   The same caveats about validation and lossiness apply as for
+--   'parsePlatformLenient'.
+--
+--   >>> parsePlatformFromSystem "x86_64-linux"
+--   Just (Platform X86_64 Linux)
+--   >>> parsePlatformFromSystem "x86_64-linux-musl"
+--   Just (Platform X86_64 Linux)
+--   >>> parsePlatformFromSystem "i686-netbsd"
+--   Just (Platform I386 NetBSD)
+parsePlatformFromSystem :: String -> Maybe Platform
+parsePlatformFromSystem s =
```

Comment:
If it has the same caveats as `parsePlatformLenient`, I’d call it `parsePlatformFromSystemLenient`

Reply:
Thanks, that's true.